### PR TITLE
8290488: IBM864 character encoding implementation bug

### PR DESCRIPTION
**Pull request by naotoj**
Adding an extra c2b mapping for the `%` in `IBM864` charset. The discrepancy came from the mapping difference between MS and IBM.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290488](https://bugs.openjdk.org/browse/JDK-8290488): IBM864 character encoding implementation bug

### Reviewers
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**)
 * [Joe Wang](https://openjdk.org/census#joehw) (@JoeWang-Java - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9661/head:pull/9661` \
`$ git checkout pull/9661`

Update a local copy of the PR: \
`$ git checkout pull/9661` \
`$ git pull https://git.openjdk.org/jdk pull/9661/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9661`

View PR using the GUI difftool: \
`$ git pr show -t 9661`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9661.diff">https://git.openjdk.org/jdk/pull/9661.diff</a>

</details>

### CODE REVIEW

**Review by irisclark — APPROVED** [review]

**Comment by takiguc** [comment]
Hello @naotoj .
I'm not reviewer, but I'd like to test this change.
Could you wait for a moment ?
Thanks.

**Comment by naotoj** [comment]
@takiguc Sure. Appreciate it.

**Comment by takiguc** [comment]
Many thanks @naotoj .

I checked the latest IBM-864 mapping table.
(I assume current OpenJDK's IBM864 may refer older mapping table)
https://raw.githubusercontent.com/unicode-org/icu/main/icu4c/source/data/mappings/ibm-864_X110-1999.ucm
.ucm file format is as follows:
https://unicode-org.github.io/icu/userguide/conversion/data.html#ucm-file-format

I checked roundtrip mapping
| IBM864.map | ibm-864_X110-1999.ucm  |
| --- | --- |
| 0x1a    U+001a | 0x1a    U+001c |
| 0x1c    U+001c | 0x1c    U+007f |
| **0x25    U+066a** | **0x25    U+0025** |
| 0x7f    U+007f | 0x7f    U+001a |
| 0x9f    U+fffd | 0x9f    U+200b |
| 0xd7    U+fec1 | 0xd7    U+fec3 |
| 0xd8    U+fec5 | 0xd8    U+fec7 |
| 0xf1    U+0651 | 0xf1    U+fe7c |

**Note**: 0x1a <-> U+001c / 0x1c <-> U+007f /  0x7f <-> U+001a entries are control character rotation for DOS.
I think it should be ignored.

I think, roundtrip side should be changed.
0x25 entry should be U+0025 on IBM864.map
Add `0x25 U+066a` into IBM864.c2b

Modify test/jdk/sun/nio/cs/mapping/Cp864.b2c for `0025 0025`
Add `0025 066a` into test/jdk/sun/nio/cs/mapping/Cp864.c2b-irreversible

This issue just for U+0025, but f possible, please add `0x9f, 0xd7, 0xd8, 0xf1` entries.

**Comment by naotoj** [comment]
Thanks for trying it out @takiguc. However, I am not planning to change any existing mappings because of the obvious compatibility issues. The fix I proposed is safe because it is additional, which used to be unmappable (thus turned into a replacement '?').

**Review by JoeWang-Java — APPROVED** [review]

**Comment by takiguc** [comment]
Hello @naotoj .

I checked [JDK-8290488](https://bugs.openjdk.org/browse/JDK-8290488).
This issue was tested by Windows 10.
I think we need to confirm expected result for b2c side to reporter.

I checked MS's 864 via following test program on my Windows 10.
```
>type b2c_1.ps1
param($code, $hex)
$h = [string]$hex
$enc_r = [Text.Encoding]::GetEncoding([int]$code)
[byte[]]$ba = @()
for($i = 0; $i -lt $h.length; $i+=2) {
  $ba += ([System.Convert]::ToInt32($h.SubString($i,2), 16))
}
$s = ""
$enc_r.GetChars($ba) | foreach {$s += [System.Convert]::ToInt32($_).ToString("X4")}
$s
>powershell -NoProfile -ExecutionPolicy Unrestricted .\b2c_1.ps1 864 25
0025
```

Please ignore about 0xD7,0xD8,0xF1 if the target platform is Windows.

Note: Test result for c2b side.
```
>type c2b_1.ps1
param($code, $hex)
$enc_r = [Text.Encoding]::GetEncoding([int]$code)
[char[]]$ca = @()
$ca += ([System.Convert]::ToInt32([string]$hex, 16))
$s = ""
$enc_r.GetBytes($ca) | foreach {$s += [System.Convert]::ToInt32($_).ToString("X2")}
$s
>powershell -NoProfile -ExecutionPolicy Unrestricted .\c2b_1.ps1 864 0025
25

>powershell -NoProfile -ExecutionPolicy Unrestricted .\c2b_1.ps1 864 066A
25
```

**Comment by naotoj** [comment]
Sorry, but I don't have any intention to change the current roundtrip behavior of 0x25 <-> U+066A in Cp/IBM864 converter. Changing it to something else will introduce an incompatible behavior with prior JDKs.

File: test/jdk/java/beans/XMLEncoder/Test4625418.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 /*
  * @test
- * @bug 4625418 8239965
+ * @bug 4625418 8239965 8290488
```
Comment (AlanBateman) [inline test/jdk/java/beans/XMLEncoder/Test4625418.java:26]: I'm curious why this XMLEncoder test is updated, is this related to round trip?

Comment (naotoj) [reply test/jdk/java/beans/XMLEncoder/Test4625418.java:26]: Yes. The test assumes all code points below U+10000 have 1-1 roundtrip (c2b2c). Since the assumption is now not met with this change, I commented out the 864 entries.

**Review by AlanBateman — APPROVED** [review]

**Comment by takiguc** [comment]
I think you can ignore my comments.
I'm not sure if this change will solve the reporter's issue...

**Comment by naotoj** [comment]
> I'm not sure if this change will solve the reporter's issue...

It does solve the submitter's issue. With this fix:
```
jshell> HexFormat.ofDelimiter(" ").formatHex("<%adc".getBytes("IBM864"))
$181 ==> "3c 25 61 64 63"
```
which is exactly the expected result.
